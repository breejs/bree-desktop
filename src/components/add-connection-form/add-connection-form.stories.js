import AddConnectionForm from './add-connection-form.vue';

const config = {
  title: 'Components/Add Connection Form/Form',
  component: AddConnectionForm
};
export default config;

const Template = (args) => ({
  components: { AddConnectionForm },
  setup() {
    return { args };
  },
  template: `<AddConnectionForm v-bind="args" />`
});

export const Default = Template.bind({});

export const Valid = Template.bind({});
Valid.args = {
  value: {
    url: 'http://localhost:8080'
  }
};

export const Invalid = Template.bind({});
Invalid.args = {
  value: {
    url: 'localhost'
  }
};
