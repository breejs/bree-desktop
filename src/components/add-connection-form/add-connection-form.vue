<script setup>
import { ref, watchEffect } from 'vue';
import ConnectionUrlInput from './connection-url-input.vue';
import ConnectionTokenInput from './connection-token-input.vue';
import ConnectionNameInput from './connection-name-input.vue';

const props = defineProps({
  /**
   * The value of the form
   * @type {object}
   * @property {string} value.name - The name of the connection
   * @property {string} value.url - The url of the connection
   * @property {string} value.token - The token of the connection
   */
  value: {
    type: Object,
    default: () => ({
      name: 'localhost',
      url: 'http://localhost:62893',
      token: ''
    })
  },
  /**
   * name of connections
   * this is passed on to the ConnectionNameInput component
   * @type {string[]}
   */
  connectionNames: {
    type: Array,
    default: () => []
  }
});

const emit = defineEmits(['submit']);

const form = ref(null);

/** maintain value of name input */
const name = ref(props.value.name ?? '');
/** maintain value of url input */
const url = ref(props.value.url ?? '');
/** maintain value of token input */
const token = ref(props.value.token ?? '');

/** name input ref */
const nameInput = ref(null);
/** url input ref */
const urlInput = ref(null);
/** token input ref */
const tokenInput = ref(null);

/**
 * form validity
 * @type {boolean}
 */
const valid = ref(false);

/**
 * submit form
 * @param {SubmitEvent} event
 */
function onSubmit(event) {
  /**
   * Triggers when form submitted
   * @property {object} value - value of connection
   * @property {string} value.name - The name of the connection
   * @property {string} value.url - url of connection
   * @property {string} value.token - token of connection
   */
  emit('submit', {
    name: name.value,
    url: url.value,
    token: token.value
  });
}

watchEffect(() => {
  // keep valid value up to date
  valid.value =
    tokenInput.value?.valid && urlInput.value?.valid && nameInput.value?.valid;
});
</script>

<template lang="pug">
button.btn.btn-success(
  type='button',
  data-bs-toggle='modal',
  data-bs-target='#add-connection-modal',
  aria-hidden='true'
)
  | Add Connection

#add-connection-modal.modal.fade(tabindex='-1', role='dialog')
  form(novalidate, ref='form', @submit.prevent='onSubmit($event)')
    .modal-dialog
      .modal-content
        .modal-header
          h4.modal-title
            | Add Connection

        .modal-body.container-fluid
          .row.g-3
            .col-12
              ConnectionNameInput(
                ref='nameInput',
                v-model='name',
                :connectionNames='connectionNames'
              )
            .col-12
              ConnectionUrlInput(ref='urlInput', v-model:value='url')
            .col-12
              ConnectionTokenInput(ref='tokenInput', v-model:value='token')

        .modal-footer
          button.btn.btn-danger.mr-3(type='button', data-bs-dismiss='modal') 
            | Cancel
          button.btn.btn-success(
            type='submit',
            :disabled='!valid',
            data-bs-dismiss='modal'
          ) 
            | Add
</template>
