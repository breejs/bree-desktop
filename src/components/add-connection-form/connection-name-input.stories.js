import ConnectionNameInput from './connection-name-input.vue';

const config = {
  title: 'Components/Add Connection Form/Connection Name Input',
  component: ConnectionNameInput
};
export default config;

const Template = (args) => ({
  components: { ConnectionNameInput },
  setup() {
    return { args };
  },
  template: '<form novalidate><ConnectionNameInput v-bind="args" /></form>'
});

export const Default = Template.bind({});

export const Valid = Template.bind({});
Valid.args = {
  value: 'valid',
  connectionNames: ['invalid']
};

export const Invalid = Template.bind({});
Invalid.args = { value: 'invalid', connectionNames: ['invalid'] };
