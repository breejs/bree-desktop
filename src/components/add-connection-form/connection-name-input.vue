<script setup>
import { onMounted, ref, watchEffect, computed } from 'vue';
import isSANB from 'is-string-and-not-blank';
import useDebounceInput from '@/composables/debounce-input';

const props = defineProps({
  connectionNames: {
    type: Array,
    default: () => []
  },
  value: {
    type: String,
    default: 'localhost'
  }
});

const emit = defineEmits(['update:value']);

const input = ref(null);

const { inputValue, valid, classes } = useDebounceInput(
  props.value,
  'localhost',
  input,
  emit,
  (value) => isSANB(value) && !props.connectionNames?.includes(value)
);

defineExpose({ valid });
</script>

<template lang="pug">
.form-floating
  input#connection-name-input.form-control(
    name='name',
    type='text',
    placeholder='',
    required,
    v-model='inputValue',
    :class='classes',
    ref='input'
  )
  label(for='connection-name-input')
    | Name*
  .invalid-feedback
    | Enter a valid name that is unique
</template>
