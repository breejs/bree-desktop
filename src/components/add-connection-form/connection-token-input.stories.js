import ConnectionTokenInput from './connection-token-input.vue';

const config = {
  title: 'Components/Add Connection Form/Connection Token Input',
  component: ConnectionTokenInput
};
export default config;

const Template = (args) => ({
  components: { ConnectionTokenInput },
  setup() {
    return { args };
  },
  template: '<form novalidate><ConnectionTokenInput v-bind="args" /></form>'
});

export const Default = Template.bind({});

export const Valid = Template.bind({});
Valid.args = {
  value:
    'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpYXQiOjE1MTYyMzkwMjJ9.tbDepxpstvGdW8TC3G8zg4B6rUYAOvfzdceoH48wgRQ'
};

export const Invalid = Template.bind({});
Invalid.args = { value: 'token' };
