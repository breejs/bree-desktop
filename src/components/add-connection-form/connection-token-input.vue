<script setup>
import { ref } from 'vue';
import useDebounceInput from '@/composables/debounce-input';

const props = defineProps({
  value: {
    type: String,
    default: ''
  }
});

const emit = defineEmits(['update:value']);

const input = ref(null);

const { inputValue, valid, classes } = useDebounceInput(
  props.value,
  '',
  input,
  emit,
  (value) => value === '' || /^[\w-=]+\.[\w-=]+\.?[\w-.+/=]*$/.test(value)
);

defineExpose({ valid });
</script>

<template lang="pug">
.form-floating
  input#connection-token-input.form-control(
    name='token',
    type='text',
    placeholder='',
    v-model='inputValue',
    :class='classes',
    ref='input'
  )
  label(for='connection-token-input')
    | Token
  .invalid-feedback
    | Enter a valid JWT token
</template>
