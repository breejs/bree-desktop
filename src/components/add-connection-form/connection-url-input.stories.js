import ConnectionUrlInput from './connection-url-input.vue';

const config = {
  title: 'Components/Add Connection Form/Connection Url Input',
  component: ConnectionUrlInput
};
export default config;

const Template = (args) => ({
  components: { ConnectionUrlInput },
  setup() {
    return { args };
  },
  template: '<form novalidate><ConnectionUrlInput v-bind="args" /></form>'
});

export const Default = Template.bind({});

export const Valid = Template.bind({});
Valid.args = { value: 'https://localhost:68' };

export const Invalid = Template.bind({});
Invalid.args = { value: 'localhost' };
