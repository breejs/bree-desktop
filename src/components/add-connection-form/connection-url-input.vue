<script setup>
import { ref, computed } from 'vue';

import useDebounceInput from '@/composables/debounce-input';

const props = defineProps({
  value: {
    type: String,
    default: 'http://localhost:62893'
  }
});

const emit = defineEmits(['update:value']);

const input = ref(null);

const { inputValue, valid, classes } = useDebounceInput(
  props.value,
  'http://localhost:62893',
  input,
  emit,
  (value) =>
    /((([A-Za-z]{3,9}:(?:\/\/)?)(?:[-;:&=+$,\w]+@)?[A-Za-z\d.-]+|(?:www\.|[-;:&=+$,\w]+@)[A-Za-z\d.-]+)((?:\/[+~%/.\w-_]*)?\??[-+=&;%@.\w_]*#?[.!/\\\w]*)?)/.test(
      value
    )
);

defineExpose({
  valid
});
</script>

<template lang="pug">
.form-floating
  input#connection-url-input.form-control(
    name='url',
    type='url',
    placeholder='http://localhost:62893',
    required,
    v-model='inputValue',
    :class='classes',
    ref='input'
  )
  label(for='connection-url-input')
    | URL*
  .invalid-feedback
    | Enter a valid URL
</template>
