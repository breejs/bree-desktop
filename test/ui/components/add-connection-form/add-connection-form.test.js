import { storyTests } from '../../_story-tests';

import * as stories from '@/components/add-connection-form/add-connection-form.stories';

storyTests(stories);
