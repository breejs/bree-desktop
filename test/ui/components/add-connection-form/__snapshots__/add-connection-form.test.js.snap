// Vitest Snapshot v1

exports[`story snapshots > Default 1`] = `
"<button class=\\"btn btn-success\\" type=\\"button\\" data-bs-toggle=\\"modal\\" data-bs-target=\\"#add-connection-modal\\" aria-hidden=\\"true\\">Add Connection</button>
<div class=\\"modal fade\\" id=\\"add-connection-modal\\" tabindex=\\"-1\\" role=\\"dialog\\">
  <form novalidate=\\"\\">
    <div class=\\"modal-dialog\\">
      <div class=\\"modal-content\\">
        <div class=\\"modal-header\\">
          <h4 class=\\"modal-title\\">Add Connection</h4>
        </div>
        <div class=\\"modal-body container-fluid\\">
          <div class=\\"row g-3\\">
            <div class=\\"col-12\\">
              <div class=\\"form-floating\\" modelvalue=\\"localhost\\"><input class=\\"form-control\\" id=\\"connection-name-input\\" name=\\"name\\" type=\\"text\\" placeholder=\\"\\" required=\\"\\"><label for=\\"connection-name-input\\">Name*</label>
                <div class=\\"invalid-feedback\\">Enter a valid name that is unique</div>
              </div>
            </div>
            <div class=\\"col-12\\">
              <div class=\\"form-floating\\"><input class=\\"form-control\\" id=\\"connection-url-input\\" name=\\"url\\" type=\\"url\\" placeholder=\\"http://localhost:62893\\" required=\\"\\"><label for=\\"connection-url-input\\">URL*</label>
                <div class=\\"invalid-feedback\\">Enter a valid URL</div>
              </div>
            </div>
            <div class=\\"col-12\\">
              <div class=\\"form-floating\\"><input class=\\"form-control\\" id=\\"connection-token-input\\" name=\\"token\\" type=\\"text\\" placeholder=\\"\\"><label for=\\"connection-token-input\\">Token</label>
                <div class=\\"invalid-feedback\\">Enter a valid JWT token</div>
              </div>
            </div>
          </div>
        </div>
        <div class=\\"modal-footer\\"><button class=\\"btn btn-danger mr-3\\" type=\\"button\\" data-bs-dismiss=\\"modal\\"> Cancel</button><button class=\\"btn btn-success\\" type=\\"submit\\" disabled=\\"\\" data-bs-dismiss=\\"modal\\"> Add</button></div>
      </div>
    </div>
  </form>
</div>"
`;

exports[`story snapshots > Invalid 1`] = `
"<button class=\\"btn btn-success\\" type=\\"button\\" data-bs-toggle=\\"modal\\" data-bs-target=\\"#add-connection-modal\\" aria-hidden=\\"true\\">Add Connection</button>
<div class=\\"modal fade\\" id=\\"add-connection-modal\\" tabindex=\\"-1\\" role=\\"dialog\\">
  <form novalidate=\\"\\">
    <div class=\\"modal-dialog\\">
      <div class=\\"modal-content\\">
        <div class=\\"modal-header\\">
          <h4 class=\\"modal-title\\">Add Connection</h4>
        </div>
        <div class=\\"modal-body container-fluid\\">
          <div class=\\"row g-3\\">
            <div class=\\"col-12\\">
              <div class=\\"form-floating\\" modelvalue=\\"\\"><input class=\\"form-control\\" id=\\"connection-name-input\\" name=\\"name\\" type=\\"text\\" placeholder=\\"\\" required=\\"\\"><label for=\\"connection-name-input\\">Name*</label>
                <div class=\\"invalid-feedback\\">Enter a valid name that is unique</div>
              </div>
            </div>
            <div class=\\"col-12\\">
              <div class=\\"form-floating\\"><input class=\\"form-control\\" id=\\"connection-url-input\\" name=\\"url\\" type=\\"url\\" placeholder=\\"http://localhost:62893\\" required=\\"\\"><label for=\\"connection-url-input\\">URL*</label>
                <div class=\\"invalid-feedback\\">Enter a valid URL</div>
              </div>
            </div>
            <div class=\\"col-12\\">
              <div class=\\"form-floating\\"><input class=\\"form-control\\" id=\\"connection-token-input\\" name=\\"token\\" type=\\"text\\" placeholder=\\"\\"><label for=\\"connection-token-input\\">Token</label>
                <div class=\\"invalid-feedback\\">Enter a valid JWT token</div>
              </div>
            </div>
          </div>
        </div>
        <div class=\\"modal-footer\\"><button class=\\"btn btn-danger mr-3\\" type=\\"button\\" data-bs-dismiss=\\"modal\\"> Cancel</button><button class=\\"btn btn-success\\" type=\\"submit\\" disabled=\\"\\" data-bs-dismiss=\\"modal\\"> Add</button></div>
      </div>
    </div>
  </form>
</div>"
`;

exports[`story snapshots > Valid 1`] = `
"<button class=\\"btn btn-success\\" type=\\"button\\" data-bs-toggle=\\"modal\\" data-bs-target=\\"#add-connection-modal\\" aria-hidden=\\"true\\">Add Connection</button>
<div class=\\"modal fade\\" id=\\"add-connection-modal\\" tabindex=\\"-1\\" role=\\"dialog\\">
  <form novalidate=\\"\\">
    <div class=\\"modal-dialog\\">
      <div class=\\"modal-content\\">
        <div class=\\"modal-header\\">
          <h4 class=\\"modal-title\\">Add Connection</h4>
        </div>
        <div class=\\"modal-body container-fluid\\">
          <div class=\\"row g-3\\">
            <div class=\\"col-12\\">
              <div class=\\"form-floating\\" modelvalue=\\"\\"><input class=\\"form-control\\" id=\\"connection-name-input\\" name=\\"name\\" type=\\"text\\" placeholder=\\"\\" required=\\"\\"><label for=\\"connection-name-input\\">Name*</label>
                <div class=\\"invalid-feedback\\">Enter a valid name that is unique</div>
              </div>
            </div>
            <div class=\\"col-12\\">
              <div class=\\"form-floating\\"><input class=\\"form-control\\" id=\\"connection-url-input\\" name=\\"url\\" type=\\"url\\" placeholder=\\"http://localhost:62893\\" required=\\"\\"><label for=\\"connection-url-input\\">URL*</label>
                <div class=\\"invalid-feedback\\">Enter a valid URL</div>
              </div>
            </div>
            <div class=\\"col-12\\">
              <div class=\\"form-floating\\"><input class=\\"form-control\\" id=\\"connection-token-input\\" name=\\"token\\" type=\\"text\\" placeholder=\\"\\"><label for=\\"connection-token-input\\">Token</label>
                <div class=\\"invalid-feedback\\">Enter a valid JWT token</div>
              </div>
            </div>
          </div>
        </div>
        <div class=\\"modal-footer\\"><button class=\\"btn btn-danger mr-3\\" type=\\"button\\" data-bs-dismiss=\\"modal\\"> Cancel</button><button class=\\"btn btn-success\\" type=\\"submit\\" disabled=\\"\\" data-bs-dismiss=\\"modal\\"> Add</button></div>
      </div>
    </div>
  </form>
</div>"
`;
